### PR TITLE
Report Full: fix delimiter line bug

### PR DESCRIPTION
**Pull request by jrfnl**
## Description
When determining the max message length, the calculation did not take potential explicit multi-line messages into account and would base the delimiter line length on the length of the complete message, not on the length of the individual lines.

Fixed now.

## Suggested changelog entry
* Report Full: fixed bug in delimiter line calculation

## Related issues/external references

Discovered while reviewing #125
Related to squizlabs/PHP_CodeSniffer#2093

## Types of changes
- [x] Bug fix _(non-breaking change which fixes an issue)_

## Additional information

Testing this fix manually is not that easy and there are no unit/integration tests for the reports yet either.

How I tested this myself is by taking the test code from squizlabs/PHP_CodeSniffer#2093 and adding it (temporarily) to a random sniff:

```php
$message = "Lorem ipsum dolor sit amet, consectetur adipiscing elit.\nAenean felis urna, dictum vitae lobortis vitae, maximus nec enim. Etiam euismod placerat efficitur. Nulla eu felis ipsum.\nCras vitae ultrices turpis. Ut consectetur ligula in justo tincidunt mattis.\n\nAliquam fermentum magna id venenatis placerat. Curabitur lobortis nulla sit amet consequat fermentum. Aenean malesuada tristique aliquam. Donec eget placerat nisl.\n\nMorbi mollis, risus vel venenatis accumsan, urna dolor faucibus risus, ut congue purus augue vel ipsum.\nCurabitur nec dolor est. Suspendisse nec quam non ligula aliquam tempus. Donec laoreet maximus leo, in eleifend odio interdum vitae.";

$phpcsFile->addWarning($message, 0, 'Test');
```

and then running it over the simple test file from #124:

```php
<?php

$x=1;
```

using the following command:
```bash
phpcs -s ./phpcs-3924.php --report-width=100000 --basepath=. --standard=universal
```
(the `universal` is because the random sniff I added the code to was in that standard)

Result when running this on `master`:
image

Result when running this against the PR branch:
image

### CODE REVIEW

File: src/Reports/Full.php [hunk 1]
```diff
@@ -69,10 +69,21 @@ public function generateFileReport($report, File $phpcsFile, $showSources=false,
                         $length += (strlen($error['source']) + 3);
                     }
 
+                    // Handle multi-line message correctly.
+                    if (strpos($error['message'], "\n") !== false) {
```
Comment (anomiex) [inline src/Reports/Full.php:73]: It seems like lines 67–70 just above could go in an `else` block to this `if`, since when this `if` is true we ignore the results of that code.

Or we could just always do the block being added here. It'll do the right thing even if there are no newlines, it'll just do a bit more work to do it. Depends how much we value speed over code complexity here, most messages shouldn't need the extra work.

**Review by anomiex — COMMENTED** [review]
Looks good to me. One suggestion inline.

File: src/Reports/Full.php (hunk 1 above)
Comment (jrfnl) [reply src/Reports/Full.php:73]: @anomiex Thanks for taking a look!

> It seems like lines 67–70 just above could go in an `else` block to this `if`, since when this `if` is true we ignore the results of that code.

The `else` would still need to be an `else if` with the `($showSources === true)` condition, so I don't see much advantage in that as it raises the cognitive load.

> Or we could just always do the block being added here. It'll do the right thing even if there are no newlines, it'll just do a bit more work to do it. Depends how much we value speed over code complexity here, most messages shouldn't need the extra work.

As message with new lines in there are very much the exception, not the rule, I'd favour performance in this case.

On that note, I do see another little tweak which can be made to remove a duplicate function call.

File: src/Reports/Full.php [hunk 2]
```diff
@@ -64,15 +64,28 @@ public function generateFileReport($report, File $phpcsFile, $showSources=false,
...
+                        $length += $srcLength;
+                    }
+
+                    // ... but also handle multi-line messages correctly.
+                    if (strpos($error['message'], "\n") !== false) {
+                        $errorLines = explode("\n", $error['message']);
+                        $length     = max(array_map('strlen', $errorLines));
+
+                        if ($showSources === true) {
+                            $lastLine = array_pop($errorLines);
```
Comment (elstamey) [inline src/Reports/Full.php:80]: It looks like we are only using the length of the last line of the error message inside this.  If `$error['message'] = "... \n ... \nthe end."` and the $showSources flag is true, we would choose the max between:

1. $length ( length of the longest line in multi-line error message ) 
2. strlen("the end.") + $srcLength

the length of the lonegest line(option 1) does not have `+ $srcLength ` included because it uses the new value of $length in line 77

File: src/Reports/Full.php (hunk 1 above)
Comment (anomiex) [reply src/Reports/Full.php:75]: It wouldn't just be an `else if`, it would be like
```php
if (strpos($error['message'], "\n") === false) {
    // Single line, calculate length.
    $length    = strlen($error['message']);

    if ($showSources === true) {
        $srcLength = (strlen($error['source']) + 3);
        $length += $srcLength;
    }
} else {
    // Multi-line, find length of longest line.
    $errorLines = explode("\n", $error['message']);
    $length     = max(array_map('strlen', $errorLines));

    if ($showSources === true) {
        $lastLine = array_pop($errorLines);
        $srcLength = (strlen($error['source']) + 3);
        $length   = max($length, (strlen($lastLine) + $srcLength));
    }
}
```
(I switched around the condition there because I find it reads better with the common one-line case first)

File: src/Reports/Full.php (hunk 2 above)
Comment (anomiex) [reply src/Reports/Full.php:80]: Let's take a slightly more concrete example,
```
$error['message'] = "First line of the message.\nSecond line of the message. This line is something long.\nthe end.";
```

What you're misunderstanding is that the output in that case would be something like
```
 0 | ERROR | First line of the message.
   |       | Second line of the message. This line is something long.
   |       | the end. (Standard.Category.Sniff.Message)
```
i.e. the source is always added to the last line.

File: src/Reports/Full.php (hunk 1 above)
Comment (anomiex) [reply src/Reports/Full.php:75]: To be clear, even if I did have any veto power here to demand changes (which I don't), I wouldn't use it on this anyway if you still feel strongly about it. I'd just approve and go.

File: src/Reports/Full.php (hunk 2 above)
Comment (jrfnl) [reply src/Reports/Full.php:80]: @elstamey Thanks for looking this over, the code, however, is correct.

This code is used to find the longest line by looping over all error messages which need to be added to the report.
The code block we're looking at is where the line length of one individual error message is determined.

The error source will only ever be added to the last line of an individual error message.

A message can either be single line, in which case, the code uses the length of the message + length of the source code as `$length`.

Or a message is multi-line, in which case `$length` is determined as follows:
* Find the longest individual line in the error message (line 75).
* Get the last line and add the error code to it, determine the length and overrule the `$length` set on line 75 if the last line including the error code would be longer than the longest individual line in the error message.

After that, the determined `$length` is compared with the the longest line length of all previously seen messages, `$maxErrorLength`, and overruling that if the longest line in the current message would be longer.

Does that make sense now ?

While writing this out, I did think of an additional efficiency tweak which could be made (`break` if `$maxErrorLength` > `$width` as there is no need to continue looking at the rest of the error messages), but I will leave that for a future PR.

Comment (elstamey) [reply src/Reports/Full.php:80]: After looking at some of the other issues and tickets linked to this one, I realized why the last line length is being grabbed when the srcLength is added.  This should have been a bit more obvious to me since the `(Squiz.Commenting.ClosingDeclarationComment.Missing)` appears at the end of the error message.  🧠 

However, I don't understand why it matters if the length of the longest line is longer than the length of the last line + $srcLength.  If #1 > #2, does the $srcLength not matter?  do we not need #1 + $srcLength in our $maxErrorLength (line 85) when we are looping?

Comment (elstamey) [reply src/Reports/Full.php:80]: OK, I am really sorry that my second comment took me so long to finish before you both responded.  Thank you so much for your examples because they helped a lot.  

For what it's worth, I did run your example test above @jrfnl and was able to see the issue and see it fixed, I just kept thinking the omission of $srcLength was going to cause a problem.  😞 

Thank you both again for taking the time to reply!

File: src/Reports/Full.php (hunk 1 above)
Comment (jrfnl) [reply src/Reports/Full.php:75]: Well, for me it would be the other way round, if the code had been submitted like that (well, without the unnecessary duplicate `$srcLength` definition), I'd approve & merge it. 😁 

When writing code myself though, I am very mindful or both performance and cognitive load/nesting levels in code.
The "how much time would it take me to grok this code if I have to look at it again in two years time ?" question.

As the single line code only contain 1 function call and 1 simple calculation which is specific to the single line code (the second function call result is used by both), the performance impact of the way the code is written now is negligible and the nesting levels and the cognitive load is lower (at least for me) 😉.

Comment (anomiex) [reply src/Reports/Full.php:75]: I think we both think either version of the code is good enough and we'd both defer to the author's preference in this situation. 🙂

Personally I don't mind the separate setting of `$srcLength` in both code paths, since only one will run and moving it out without potentially setting it unnecessarily would mean having to duplicate the `if ($showSources === true) {` check. Then again, if I were writing it entirely myself I'd probably not have `$srcLength` at all as I'd find inlining the `strlen` call into the expressions easier to follow.

Getting further off topic: I've always found judging cognitive load tricky, I've had plenty of experience both with code I find easy to understand that others tell me they find hard and code I find hard to follow that others tell me they find easier. I've never been able to get a grasp on to what extent that's because I personally think differently from most people and to what extent it's that everyone thinks differently from others.

Comment (jrfnl) [reply src/Reports/Full.php:75]: I hear you. Let's just say that for now, this patch has had enough love & attention and the code can go in. A future iteration (maybe when I work on the tests), can sort out other tweaks and optimizations.

Thanks for reviewing!

**Comment by jrfnl** [comment]
Rebased & squashed without further changes. I will merge this once the build has passed.
